src/main/java/io/cdap/plugin/spark/HierarchyToRelationalUtils.java

```diff
@@ -0,0 +1,224 @@
+      structuredRecord -> DataFrames.toRow(structuredRecord, DataFrames.toDataType(inputSchema)));
+    SQLContext sqlContext = new SQLContext(sparkExecutionPluginContext.getSparkContext());
+    Dataset<Row> data = sqlContext.createDataFrame(map, DataFrames.toDataType(inputSchema));
+    data.registerTempTable(SQLUtils.NAME_HIERARCHY_TABLE);
+    Dataset<Row> rootRecords = sqlContext.sql(SQLUtils.sqlStatementForRoot(config, SQLUtils.NAME_HIERARCHY_TABLE));
+    if (rootRecords.select(config.getParentField()).distinct().count() > 1) {
+      throw new Exception("There can only be one root element in the hierarchy");
+    }
+    Row root = rootRecords.select(config.getParentField()).distinct().first();
+
+    Dataset<Row> uniqueRecords = data.sqlContext().sql(
+      SQLUtils.sqlForUniqueRecords(config, root, SQLUtils.NAME_HIERARCHY_TABLE));
+    // Fetch on child id's from the input dataset
+    Dataset<Row> leafRecords = data.sqlContext().sql(
+      SQLUtils.sqlStatementForLeafRecords(config, SQLUtils.NAME_HIERARCHY_TABLE));
+    leafRecords.registerTempTable(SQLUtils.NAME_LEAVES_TEMP_TABLE);
+    // NOTE - required in order to be able to iterate - can't pass data and sqlContext otherwise
+    List<Row> rows = uniqueRecords.collectAsList();
+
+    // Empty Dataset to hold all the results
+    Dataset<Row> results = sqlContext.createDataFrame(
+      sparkExecutionPluginContext.getSparkContext().emptyRDD(), DataFrames.toDataType(outputSchema));
+    // Root does not exists as separate value - this will generate it
+    results = results.union(generateTreeRoot(data, config, nonMappedFields, mappedFields));
+    // Generate tree for each unique row
+    for (Row row : rows) {
+      results = results.union(generateQuery(row, data, sqlContext));
+    }
+    // sets leaves indicator
+    results.registerTempTable(SQLUtils.NAME_RESULTS_TEMP_TABLE);
+    return sqlContext.sql(SQLUtils.sqlStatementForSettingLeavesValue(config, nonMappedFields, mappedFields));
+  }
+
+  /**
+   * Generate dataset containing all subtree details for a given row
+   *
+   * @param parentRow  row for which to extract the subtree
+   * @param data       reference to input dataset
+   * @param sqlContext reference to {@link SQLContext}
+   * @return dataset reference with all subtree rows for a given row
+   * @throws Exception raises exception if max depth is exceeded
+   */
+  public Dataset<Row> generateQuery(Row parentRow, Dataset<Row> data, SQLContext sqlContext)
+    throws Exception {
+    long dataFrameCount = 1;
+    int count = 1;
+
+    // make parent id sql ready based on type
+    String parentId;
+    String parentIdValue = String.valueOf(parentRow.get(0));
+    if (NumberUtils.isNumber(parentIdValue)) {
+      parentId = parentIdValue;
+    } else {
+      parentId = String.format("'%s'", parentRow.getString(0));
+    }
+
+    mapSelfToTreeRoot(parentRow, data, config);
+    List<String> childProductFields = new ArrayList<>(mappedFields.values());
```

Comment:
Rename variables childProductFields, parentProductFields. The names "ParentProduct", "ChildProduct" are not hardcoded.